### PR TITLE
feat: save all already processed events

### PR DESCRIPTION
**Pull request by marcellorigotti**
# Pull Request

Closes: PRO-2162
## Checklist

Please conduct a thorough self-review before opening the PR.

- [x] I am confident that the code works.
- [x] I have written sufficient tests.
- [ ] I have written and tested required migrations.
- [ ] I have updated documentation where appropriate.

## Summary

In order to handle some edge cases we changed the approach:
- we now save all processed events until the safety margin expires, in case of duplicated we always save the highest expiry_block (block_number + safety margin)

### CODE REVIEW

File: state-chain/pallets/cf-elections/src/electoral_systems/block_witnesser/block_processor.rs [hunk 1]
```diff
@@ -203,57 +181,62 @@ impl<T: BWProcessorTypes> BlockProcessor<T> {
...
+						if *stored_expiry < expiry {
+							*stored_expiry = expiry;
+						}
+						false
+					},
+					None => {
+						self.processed_events.insert(event.clone(), expiry);
+						true
+					},
+				}
```
Comment (MxmUrw) [inline state-chain/pallets/cf-elections/src/electoral_systems/block_witnesser/block_processor.rs:231]: It's a bit surprising that this function mutates state, because it feels like one might to want to call it without the mutation effect. If that's not true then I'm fine with keeping it this way, though I would try to update the name of the function to make clear that it is mutating / has side effects.

Thoughts @kylezs ?

File: state-chain/pallets/cf-elections/src/electoral_systems/block_witnesser/block_processor.rs [hunk 2]
```diff
@@ -203,57 +181,62 @@ impl<T: BWProcessorTypes> BlockProcessor<T> {
...
 		events
 			.into_iter()
-			.filter(|(_, last_event)| {
-				!self
-					.reorg_events
-					.iter()
-					.flat_map(|(_, (events, _))| events)
-					.any(|event| event == last_event)
+			.filter(|(block_number, event)| {
+				let expiry = block_number.saturating_forward(safety_margin as usize);
```
Comment (MxmUrw) [inline state-chain/pallets/cf-elections/src/electoral_systems/block_witnesser/block_processor.rs:219]: Nice, storing the expiry directly makes things definitely more understandable!

Can we, instead of using `current_block_number + safety` as expiry, use `max{all blocks we know, "end of current range"}`. That would make it a bit more resistant to bad behaved reorgs. We could compute that number in `process_block_data` and pass it in to this function I think.

**Review by MxmUrw — COMMENTED** [review]
Thanks, looks good, left some comments!

**Comment by marcellorigotti** [comment]
Closing this one, will open it against the new branch `feat/block-witnesser-es-rebase` which is rebased against main
